src/core/api.c

```diff
@@ -1615,6 +1615,102 @@ MsQuicGetParam(
     return Status;
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QUIC_API
+MsQuicHandleAddRef(
+    _In_ _Pre_defensive_
+        HQUIC Handle
+    )
+{
+    CXPLAT_PASSIVE_CODE();
+
+    QuicTraceEvent(
+        ApiEnter,
+        "[ api] Enter %u (%p).",
+        QUIC_TRACE_API_ADDREF,
+        Handle);
+
+    if (IS_CONN_HANDLE(Handle)) {
+        #pragma prefast(suppress: __WARNING_25024, "Pointer cast already validated.")
+        QUIC_CONNECTION* Connection = (QUIC_CONNECTION*)Handle;
+
+        QUIC_CONN_VERIFY(Connection, !Connection->State.HandleClosed);
+        QUIC_CONN_VERIFY(Connection, !Connection->State.Freed);
+
+        InterlockedIncrement((volatile long*)&Connection->ExternalRefCount);
+    } else if (IS_STREAM_HANDLE(Handle)) {
+        #pragma prefast(suppress: __WARNING_25024, "Pointer cast already validated.")
+        QUIC_STREAM* Stream = (QUIC_STREAM*)Handle;
+
+        CXPLAT_TEL_ASSERT(!Stream->Flags.HandleClosed);
+        CXPLAT_TEL_ASSERT(!Stream->Flags.Freed);
+
+        InterlockedIncrement((volatile long*)&Stream->ExternalRefCount);
+    } else {
+        CXPLAT_FRE_ASSERTMSG(FALSE, "Only supported for Connection and Stream handles!");
+    }
+
+    QuicTraceEvent(
+        ApiExit,
+        "[ api] Exit");
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+BOOLEAN
+QUIC_API
+MsQuicHandleRelease(
+    _In_ _Pre_defensive_
+        HQUIC Handle
+    )
+{
+    BOOLEAN WasClosed = FALSE;
+
+    CXPLAT_PASSIVE_CODE();
+
+    QuicTraceEvent(
+        ApiEnter,
+        "[ api] Enter %u (%p).",
+        QUIC_TRACE_API_RELEASE,
+        Handle);
+
+    if (IS_CONN_HANDLE(Handle)) {
+        #pragma prefast(suppress: __WARNING_25024, "Pointer cast already validated.")
+        QUIC_CONNECTION* Connection = (QUIC_CONNECTION*)Handle;
+
+        QUIC_CONN_VERIFY(Connection, !Connection->State.HandleClosed);
+        QUIC_CONN_VERIFY(Connection, !Connection->State.Freed);
+
+        long RefCount = InterlockedDecrement((volatile long*)&Connection->ExternalRefCount);
+
+        if (RefCount <= 0) {
```

Comment:
Catch double release calls. If RefCount goes negative, FRE_ASSERT

Reply:
We don't want to FRE_ASSERT here. We want to do the same DBG_ASSERTs changes that we did for close. The easiest way to do that is to just forward to close a 2nd time.

Reply:
Why would we not want to FRE_ASSERT? This is essentially a double free?

Reply:
Its a double free that isn't actually free'd yet, so all the memory is perfectly valid. Its kind of an annoying side affect of how our shutdown model works.

https://github.com/microsoft/msquic/pull/2239

Its using the change in the above PR to make this work.

Reply:
How do we know the object hasn't been free'd yet? I don't think there's any guarantee of that? Also calling Close twice could double-free other objects. Possibly not today, but in the future?

Reply:
We know the object hasn't actually been freed because by ref count rules the only way to get a reference is on the worker thread. And if you have a reference and are on the worker thread, it has not been freed. This is UB if the caller calls Release again from the non worker thread, but by then the ref count has been freed anyway, so its UB to even do the decrementing. Theres nothing to FRE_ASSERT that isn't UB in that case.

By just forwarding this to Close, we only have to ensure Close keeps the current semantics of not hard failing if the app makes a mistake. As long as we don't change Close to double free anything without checking if we're in the worker, we're good.